### PR TITLE
ArchiveSpaceManager Implementation

### LINKED ISSUE #162: Module for maintaining diskspace
**Is your feature request related to a problem? Please describe.**
typical NVR applications are expected to watch the available disk space and make sure that the NVR does not crash due to disk space full errors hence it is important to have a module which can watch the archieve 's disk utilization

**Describe the solution you'd like**
A module which is able to watch the disk space and clean old files. Following simple props apply

1. Path: the pathe to watch (e.g. /var/nvr/cam1)
2. upper_water_mark: space in MB (e.g. 900MB) which triggers a corrective actions (e.g. deletion)
3. lower_water_mark:  space in MB (e.g. 750MB) which the a corrective action aspires to acheive 
4. a pattern for file name which can be deleted to acheive the move from high_water_mark to lower_water_mark (e.g. *.mp4)

**Describe alternatives you've considered**
there is linux utility named "tmpreaper" that can be considered here

**Additional context**

we should try to avoid making too many filesystem queries in learning the current state... a good approach can be as follows

- At a very low frequency we do a fine analysis, while at a higher (but still low) frequency we do coarser analysis
- fine analysis => we know files sizes of each and every file
- coarse analysis => just count the additional files and project based on some averaging scheme (average size of latest 10 files can be used to predict the size of next 10 files)
- goal is to defer the fine analysis as mush as possible

**Additionally** strive to avoid disk full situation (other than the high and low water mark)
e.g. if the high watermark is misconfigured to be more than available disk space... we should try to avoid making diskful and go to lower water mark

### PR DESCRIPTION
**Pull request by Vinayak-YB**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #162 

**Description**

Archive space manager implementation. The module takes upperWaterMark, lowerWaterMark, path as properties. If the archive space occupied is greater than upperWaterMark then oldest files are deleted till lowerWaterMark is reached. 

**Type**

Type Choose one: Feature

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/src/ArchiveSpaceManager.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        return tempSize;
+    }
+    ArchiveSpaceManagerProps mProps;
+    uint32_t archiveSize = 0;
+    std::vector<std::pair<boost::filesystem::path, uint32_t>> foldVector;
+};
+
+
+ArchiveSpaceManager::ArchiveSpaceManager(ArchiveSpaceManagerProps _props)
+    :Module(TRANSFORM, "DiskspaceManager", _props)
```
Comment (kumaakh) [inline base/src/ArchiveSpaceManager.cpp:111]: Kind shoud be SOURCE
The space manager is a free running thread which decides to exampine disk space periodically

File: base/test/archivespacemanager_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+};
+
+BOOST_AUTO_TEST_CASE(basic)
+{
+	createDiskFiles();
+	std::vector<std::string> folder = { "./data/archive" };
+	Test_Utils::FileCleaner f(folder);
+	uint32_t lowerWaterMark = 25000;
+	uint32_t upperWaterMark = 30000;
+	auto source = boost::shared_ptr<ExternalSourceModule>(new ExternalSourceModule());
```
Comment (kumaakh) [inline base/test/archivespacemanager_tests.cpp:47]: we dont need an external source module to decide whien to trigger it... it runs on it's own.

File: base/src/ArchiveSpaceManager.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    uint32_t diskOperation()
+    {
+        archiveSize = estimateDirectorySize(mProps.pathToWatch);
+        if (archiveSize > mProps.upperWaterMark)
+        {
+            manageDirectory();
+        }
+        uint32_t tempSize = archiveSize;
+        archiveSize = 0;
```
Comment (kumaakh) [inline base/src/ArchiveSpaceManager.cpp:101]: why do we need to set archievesize to 0 ?

File: base/src/ArchiveSpaceManager.cpp [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            {
+                if (!(countFreq % mProps.samplingFreq))
+                {
+                    sample = (rand() % mProps.samplingFreq) + 1;
+                    inCount = 0;
+                }
+                if (inCount == sample)
+                {
+                    tempSize = boost::filesystem::file_size(entry);
+                    dirSize += tempSize * mProps.samplingFreq;
```
Comment (kumaakh) [inline base/src/ArchiveSpaceManager.cpp:43]: what if the total number of files is not a multipe of mProps.samplingFreq? e.g. what if you have 46-59 files in an hour ? the last batch now has 1-14 files while you are multiplying by 15

File: base/include/ArchiveSpaceManager.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
+#pragma once
+#include "Module.h"
+#include "AIPExceptions.h"
+
+class ArchiveSpaceManagerProps : public ModuleProps
+{
+public:
+	ArchiveSpaceManagerProps(uint32_t _lowerWaterMark, uint32_t _upperWaterMark, string _pathToWatch, int _samplingFreq)
```
Comment (kumaakh) [inline base/include/ArchiveSpaceManager.h:8]: we should support another construcotr which just takes max_size_allowed and computes lower water mark from their as 10% less

File: base/include/ArchiveSpaceManager.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	{
+		lowerWaterMark = _lowerWaterMark;
+		upperWaterMark = _upperWaterMark;
+		pathToWatch = _pathToWatch; 
+		samplingFreq = _samplingFreq;
+		
+		auto totalSpace = boost::filesystem::space(pathToWatch);
+		if ((lowerWaterMark > upperWaterMark) || (upperWaterMark > totalSpace.available))
```
Comment (kumaakh) [inline base/include/ArchiveSpaceManager.h:16]: upperWaterMark can be greater than totalSpaceAvailable! this check should be done on the capacity of the disk, the same space function provides capacity, free and available

File: base/src/ArchiveSpaceManager.cpp [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                foldVector.push_back({oldHrDir,boost::filesystem::last_write_time(oldHrDir) });
+            }
+            sort(foldVector.begin(), foldVector.end(), comparator); //Sorting the vector
+
+            uint32_t tempSize = 0;
+            boost::filesystem::path delDir = foldVector[0].first;
+            BOOST_LOG_TRIVIAL(info) <<"Deleting file : "<< delDir.string();
+            tempSize = estimateDirectorySize(delDir);
+            archiveSize = archiveSize - tempSize;
+            boost::filesystem::remove_all(delDir);
```
Comment (kumaakh) [inline base/src/ArchiveSpaceManager.cpp:88]: we should handle any delete exception here.

File: base/include/ArchiveSpaceManager.h [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        {
+			LOG_ERROR << "Please enter correct properties!";
+			std::string errorMsg = "Incorrect properties set for Archive Manager. TotalDiskSpace <" + std::to_string(totalSpace.available) + ">lowerWaterMark<" + std::to_string(lowerWaterMark) + "> UpperWaterMark<" + std::to_string(upperWaterMark) + ">";
+			throw AIPException(AIP_FATAL,errorMsg );
+        }
+	}
+
+	uint32_t lowerWaterMark = 1000000000; // Lower disk space
```
Comment (kumaakh) [inline base/include/ArchiveSpaceManager.h:24]: why do we need these default values ? we dont have a default consturctor

**Review by kumaakh — CHANGES_REQUESTED** [review]

File: base/src/ArchiveSpaceManager.cpp (hunk 1 above)
Comment (Vinayak-YB) [reply base/src/ArchiveSpaceManager.cpp]: ArchiveSpaceManager is changed to source and test cases are refactored.

File: base/test/archivespacemanager_tests.cpp (hunk 2 above)
Comment (Vinayak-YB) [reply base/test/archivespacemanager_tests.cpp]: External Source module is removed and ArchiveSpaceManager is made source.

File: base/src/ArchiveSpaceManager.cpp (hunk 4 above)
Comment (Vinayak-YB) [reply base/src/ArchiveSpaceManager.cpp:43]: The no of files in last batch are multiplied with tempSize and added to dirSize after for loop.

File: base/include/ArchiveSpaceManager.h (hunk 5 above)
Comment (Vinayak-YB) [reply base/include/ArchiveSpaceManager.h:8]: A constructor which takes max_size and computes lower water mark (10% less) is added.

File: base/src/ArchiveSpaceManager.cpp (hunk 7 above)
Comment (Vinayak-YB) [reply base/src/ArchiveSpaceManager.cpp]: Try() Catch(..) is added in the delete logic.

File: base/src/ArchiveSpaceManager.cpp (hunk 3 above)
Comment (Vinayak-YB) [reply base/src/ArchiveSpaceManager.cpp:108]: The archivesize is reset after complete delete operation

**Review by kumaakh — CHANGES_REQUESTED** [review]
Lets change the fps in ModuleProps to be a "float" rather than "int"
and make sure that "ArchiveSpaceManagerProps" constructors set it to 0.001. this way if the developer forgets to initial it we will not have a bad effect

**Review by kumaakh — APPROVED** [review]

**Pull request merged**